### PR TITLE
Penalise variation of beta only where ice is grounded AND there is available data.

### PR DESCRIPTION
**Pull request by dngoldberg**
This change addresses a bug relating to prior.py. Addresses [Issue 14](https://github.com/EdiGlacUQ/fenics_ice/issues/14)

Existing Behavior:

prior.Laplacian_flt.prior_form makes use of solver.float_conditional to define the prior functional/distribution, penalising (with weight delta_beta_gnd) variation between beta and an imposed function for beta (based on temperature-derived ice stiffness Bglen/Bbar) -- lines 335-337. This is applied wherever ice is grounded. The imported beta field (also used as initial guess for beta) is currently extrapolated from the coverage of a larger externally produced data product. This means that where ice is grounded, but the external data product is undefined, this condition is still applied. 

The modification consists of:
1. Changing model.bglen_from_data to read in 2 fields: one is an extrapolated product to generate initial guess for beta. Other ("Bglenmask") is a field which is NaN where the original prodocut is not provided; this is interpolated to a DG(0) function and set to 0 where nan, 1 otherwise. (changes made to config.py and inout.py to enable this.)
2. Define a new conditional solver.bglen_data_conditional that is true where (a) ice is grounded and (b) a data constraint on beta exists.
3. Modify prior.Laplacian_flt.prior_form to use the difference between beta and beta_bgd only where this new conditional is true. beta_bgd is THE SAME as before this fix, but should be "correct" as the constraint is only applied where the external field is available. 

NOTE: this approach means there could be DoFs where the values used to constrain in (3) are in fact derived from extrapolation on the original product grid. A safer way to do this would be to read "Bglenmask" from the .h5 file as a CG(1) function, determine all elements where one vertex is nan, and define a new DG(0) function that is 1 only where all 3 vertices are non-nan. But I do not know how to do this. This result would replace model.bglen_mask.

NOTE: Smith_glacier repository must be updated to create 2 .h5 files (or an .h5 file with 2 fields). I will lead this.

### CODE REVIEW

**Comment by bearecinos** [comment]
Hi @dngoldberg looking at this I don't think it will bring conflicts I forgot to say that you can also test the taylor verification locally by running this command:
``pytest -m tv``

This will run only the taylor verification test.

File: fenics_ice/model.py [hunk 1]
```diff
@@ -148,7 +148,17 @@ def alpha_from_data(self):
     def bglen_from_data(self):
         """Get bglen field from initial input data"""
         self.bglen = self.input_data.interpolate("Bglen", self.Q)
-
+        """Get bglen mask field from input data"""
+        self.bglen_mask = self.input_data.interpolate("Bglenmask", self.M)
```
Comment (jrmaddison) [inline fenics_ice/model.py:152]: Can the `interpolate` method be used safely with values that may be NaN? If not we should first convert to zero/non-zero before interpolation.

File: fenics_ice/model.py [hunk 2]
```diff
@@ -148,7 +148,17 @@ def alpha_from_data(self):
...
+
+        """bglen_mask is currently just bglen where data available and nan elsehere"""
+        """Following is hacky way to convert this to 0 where nan, 1 elsewhere"""
+        temp_vec = self.bglen_mask.vector()[:]
```
Comment (jrmaddison) [inline fenics_ice/model.py:156]: `self.bglen_mask.vector().get_local()` here, or can use the tlm_adjoint interface, `temp_vec = function_get_local(self.bglen_mask)`

File: fenics_ice/model.py [hunk 3]
```diff
@@ -148,7 +148,17 @@ def alpha_from_data(self):
...
+        temp_vec[np.isnan(temp_vec)] = 0.0
+        temp_vec[temp_vec>0.0]=1.0
```
Comment (jrmaddison) [inline fenics_ice/model.py:158]: `temp_vec = np.where(np.isnan(temp_vec), 1.0, 0.0)`

File: fenics_ice/model.py [hunk 4]
```diff
@@ -148,7 +148,17 @@ def alpha_from_data(self):
...
+        self.bglen_mask.vector()[:] = temp_vec
+        """Following appears in inout.interpolate -- is something similar needed?"""
+        self.bglen_mask.vector().apply("insert")
```
Comment (jrmaddison) [inline fenics_ice/model.py:161]:
```
self.bglen_mask.set_local(temp_vec)
self.bglen_mask.vector().apply("insert")
```
or can use the tlm_adjoint interface, `function_set_local(self.bglen_mask, temp_vec)`.

File: fenics_ice/solver.py [hunk 5]
```diff
@@ -1063,6 +1070,29 @@ def float_conditional(self, H, H_float=None):
 
         return fl_ex
 
+    def bglen_data_conditional(self, H, glen_mask, H_float=None):
+        """Compute a ufl Conditional where floating=1, grounded=0"""
+
+        if not self.params.ice_dynamics.allow_flotation:
+            fl_beta_mask = ufl.operators.Conditional(self.model.bglen_mask == 1, 
+                                          Constant(1.0, cell=triangle, name="Const data"),
+                                          Constant(0.0, cell=triangle, name="Const no data"))
```
Comment (jrmaddison) [inline fenics_ice/solver.py:1079]: `> 0.0` in the conditional, for consistency with below.

File: fenics_ice/model.py (hunk 1 above)
Comment (bearecinos) [reply fenics_ice/model.py:152]: as long as I am aware these two functions [interpolate()](https://github.com/EdiGlacUQ/fenics_ice/blob/3c63d1fe968aeebbce8d0ac10a1af578b768f4df/fenics_ice/model.py#L232-L257) and the one starting on L232 do not handle nan's as stated in the comments.
```
def interpolate(values, vtx, wts):
            """Bilinear interpolation, given vertices & weights above"""
            return np.einsum('nj,nj->n', np.take(values, vtx), wts)
```

Comment (dngoldberg) [reply fenics_ice/model.py:152]: I wanted to ensure that any cell that could possibly be affected would have its DoF as Nan. @bearecinos i believe this interpolate function is defined in inout.py

File: fenics_ice/model.py (hunk 4 above)
Comment (dngoldberg) [reply fenics_ice/model.py:161]: @jrmaddison  see new changes to this function. i still use the "insert" argument for `apply`, is this appropriate?

File: fenics_ice/model.py (hunk 3 above)
Comment (dngoldberg) [reply fenics_ice/model.py:158]: @jrmaddison  see new changes to this function

File: fenics_ice/model.py (hunk 2 above)
Comment (dngoldberg) [reply fenics_ice/model.py:156]: @jrmaddison  see new changes to this function

File: fenics_ice/model.py (hunk 1 above)
Comment (dngoldberg) [reply fenics_ice/model.py:152]: @jrmaddison  see new changes to this function

File: fenics_ice/solver.py (hunk 5 above)
Comment (dngoldberg) [reply fenics_ice/solver.py:1079]: have modified the 2nd conditional to `bglen_mask == 1` instead

File: fenics_ice/model.py (hunk 4 above)
Comment (jrmaddison) [reply fenics_ice/model.py]: I'm not sure `apply` is needed here, but it seems safer to leave it.

**Comment by dngoldberg** [comment]
@jrmaddison  are you satisfied with changes so that i can merge?

File: fenics_ice/model.py [hunk 6]
```diff
@@ -148,7 +148,28 @@ def alpha_from_data(self):
...
+        temp_vec = bglen_mask_CG.vector().gather(np.arange(self.Q.dim()))
+        bmask_loc = np.ones(self.bglen_mask.vector().local_size(), dtype=np.float64)
+        for cell in range(self.mesh.num_cells()):
+          if not Cell(self.mesh, cell).is_ghost():
+            local_dofs = dofmap_Q.cell_dofs(cell)
+            global_dofs = np.empty(np.shape(local_dofs),dtype=int)
+            for dof in range(len(local_dofs)):
+              global_dofs[dof] = dofmap_Q.local_to_global_index(local_dofs[dof])
+            if (temp_vec[global_dofs] < 1.0-1.0e-10).any():
+              bmask_loc[dofmap_M.cell_dofs(cell)] = 0
```
Comment (jrmaddison) [inline fenics_ice/model.py:168]: Fine if `self.M` is discontinuous Lagrange.

**Review by jrmaddison — APPROVED** [review]

File: fenics_ice/model.py [hunk 7]
```diff
@@ -148,7 +148,28 @@ def alpha_from_data(self):
...
-
+        """Get bglen mask field from input data"""
+        bglen_mask_CG = self.input_data.interpolate("Bglenmask", self.Q)
+        self.bglen_mask = Function(self.M, name="bglen_mask")
+
+        dofmap_M = self.M.dofmap()
+        dofmap_Q = self.Q.dofmap()
+
+        """bglen_mask is currently 1 where there is data, zero elsewhere"""
+        temp_vec = bglen_mask_CG.vector().gather(np.arange(self.Q.dim()))
```
[the remark this answers is not in the document]
Comment (dngoldberg) [reply fenics_ice/model.py:159]: yes this is what was intended, and am fine with it..

File: fenics_ice/model.py [hunk 8]
```diff
@@ -148,7 +148,28 @@ def alpha_from_data(self):
...
+        bmask_loc = np.ones(self.bglen_mask.vector().local_size(), dtype=np.float64)
+        for cell in range(self.mesh.num_cells()):
+          if not Cell(self.mesh, cell).is_ghost():
+            local_dofs = dofmap_Q.cell_dofs(cell)
+            global_dofs = np.empty(np.shape(local_dofs),dtype=int)
```
[the remark this answers is not in the document]
Comment (dngoldberg) [reply fenics_ice/model.py:164]: thank you.

**Pull request merged**